docs/src/lib/functions.md

```diff
@@ -48,4 +48,62 @@ sort!
 tail
 unique!
 permutecols!
-```
\ No newline at end of file
+```
+
+## Indexing
+
+The following rules explain target functionality of how `getindex` and `view` are intended to work for `DataFrame`, `SubDataFrame` and `DataFrameRow`.
+
+The rules for a valid type of index into a column are the following:
+* a value, later denoted as `col`:
+    * a `Symbol`;
+    * an `Integer` that is not `Bool`;
+* a vector, later denoted as `cols`:
+    * a vector of `Symbol` (does not have to be a subtype of `AbstractVector{Symbol}`);
+    * a vector of `Integer` other than `Bool` (does not have to be a subtype of `AbstractVector{<:Integer}`);
+    * a vector of `Bool` that has to be a subtype of `AbstractVector{Bool}`.
+    * a colon.
+
+The rules for a valid type of index into a row are the following:
+* a value, later denoted as `row`:
+    * an `Integer` that is not `Bool`;
+* a value, later denoted as `rows`:
+    * a vector of `Integer` other than `Bool` (does not have to be a subtype of `AbstractVector{<:Integer}`);
+    * a vector of `Bool` that has to be a subtype of `AbstractVector{Bool}`;
+    * a colon.
+
+The following list specifies return types of `getindex` operations depending on argument types. In the summary `df` represents a `DataFrame`, `sdf` is a `SubDataFrame` and `dfr` is a `DataFrameRow`.
+
+`DataFrame`:
+* `df[col]` -> an alias of a vector contained in column `col`;
+* `df[cols]` -> a freshly allocated `DataFrame` containing aliases of vectors contained in columns `cols`;
+* `df[row, col]` -> a value contained in row `row` of column `col`;
+* `df[row, cols]` -> a `NamedTuple` containing data from row `row` in columns `cols`;
+* `df[rows, col]` -> a copy of a vector `df[col]` with only rows `rows` selected;
+* `df[rows, cols]` -> a `DataFrame` containing columns `cols` and `df[rows, col]` as a vector in each `col` in `cols`.
+* `@view df[col]` -> an alias of a vector contained in column `col`;
+* `@view df[cols]` -> a `SubDataFrame` with parent `df` if `cols` is a colon and `df[cols]` otherwise;
+* `@view df[row, col]` -> translates to `view(df[col], row)` (a `0`-dimensional view into `df[col]`);
+* `@view df[row, cols]` -> a `DataFrameRow` with parent `df` if `cols` is a colon and `df[cols]` otherwise;
+* `@view df[rows, col]` -> translates to `view(df[col], rows)` (a standard view into `df[col]` vector);
+* `@view df[rows, cols]` -> a `SubDataFrame` with parent `df` if `cols` is a colon and `df[cols]` otherwise.
+
+`SubDataFrame`:
+* `sdf[col]` -> a view of vector contained in column `col` of `parent(sdf)` with `DataFrames.rows(sdf)` as selector;
+* `sdf[cols]` -> a freshly allocated `DataFrame` containing copies of vectors contained in columns `cols` of `sdf`;
```

Comment:
This seems inconsistent with `sdf[col]` and `df[cols]`, which don't make copies. Better return a `SubDataFrame`.

Reply:
I was not sure what is best here myself.

In Base such an operation destroys a view. Also some code in DataFrames.jl relies on the fact that `sdf[cols]` returns a `DataFrame` if I remember correctly.

I will leave as is in the next iteration, but let us discuss it further.

Reply:
But in Base `x[inds]` always makes a copy, so it's consistent that it's also the case for views. For `DataFrame`, `df[cols]` doesn't copy the column vectors, so it would be consistent to do that for `SubDataFrame`. The goal is that for consistency if you mutate the result, the original `AbstractDataFrame` is changed in both cases.

Reply:
OK, I have rewritten it to:
> a `SubDataFrame`, with parent `parent(sdf)` if `cols` is a colon and `parent(sdf)[cols]` otherwise